src/block/decompress.rs

```diff
@@ -1,43 +1,91 @@
 //! The decompression algorithm.
-use crate::block::wild_copy_from_src_16;
 use crate::block::DecompressError;
 use crate::block::Sink;
+use crate::block::MINMATCH;
 use alloc::vec::Vec;
 
-// copy_on_self uses 16byte wild copy, to avoid overlapping copies we add 12. Minimum length of match_length is 4 totaling to 16.
-const SAFE_DUPLICATE_COPY_RANGE: usize = 12;
-
 /// Copies data to output_ptr by self-referential copy from start and match_length
 #[inline]
-fn duplicate(output_ptr: &mut *mut u8, start: *const u8, match_length: usize) {
+unsafe fn duplicate(
+    output_ptr: &mut *mut u8,
+    output_end: *mut u8,
+    start: *const u8,
+    match_length: usize,
+) {
     // We cannot simply use memcpy or `extend_from_slice`, because these do not allow
     // self-referential copies: http://ticki.github.io/img/lz4_runs_encoding_diagram.svg
-    // Check overlap copy
 
-    if (*output_ptr as usize) < start as usize + match_length + SAFE_DUPLICATE_COPY_RANGE {
+    // Considering that `wild_copy_match_16` can copy up to `16 - 1` extra bytes.
+    // Defer to `duplicate_overlapping` in case of an overlapping match
+    // OR the if the wild copy would copy beyond the end of the output.
+    if start.add(match_length + 16 - 1) > *output_ptr
+        || output_ptr.add(match_length + 16 - 1) > output_end
```

Comment:
I'm currently checking a performance regression for unsafe decompression (66K JSON 5GB/s down to 4GB/s). It is mainly caused by this extended check

Reply:
That's odd, 1Gbp/s from this branch in the slow path? The compiler is going crazy.

Reply:
This commit comes to mind https://github.com/PSeitz/lz4_flex/pull/13/commits/0a158955e4ff742a5b494c23d87a3d397c1661c4

But it's in a different location.

Reply:
Something that can help a lot is to run both versions alongside, in same binary.
The bigger benchmark binaries now can throw LTO offtrack significantly, it's almost random sometimes.

```toml
[dev-dependencies.lz4_flex_master]
git = "https://github.com/PSeitz/lz4_flex"
rev= "eb8c2e090485dcb5ef9e0da96bf72a95023753c1" # before this PR was merged
package = "lz4_flex"
```

Then it's available as `lz4_flex_master` in the benchmark code.

Reply:
I tried the above and this branch is 15% faster so something might be affecting your comparison.

```
BlockDecompress/lz4_flex_rust_master/66675                                                                             
                        time:   [10.145 us 10.153 us 10.162 us]
                        thrpt:  [6.1108 GiB/s 6.1161 GiB/s 6.1208 GiB/s]
                 change:
                        time:   [-2.2150% -2.1219% -2.0144%] (p = 0.00 < 0.05)
                        thrpt:  [+2.0558% +2.1679% +2.2652%]
                        Performance has improved.
Found 2 outliers among 100 measurements (2.00%)
  2 (2.00%) high mild
```
```
BlockDecompress/lz4_flex_rust/66675                                                                             
                        time:   [8.7167 us 8.7213 us 8.7266 us]
                        thrpt:  [7.1157 GiB/s 7.1200 GiB/s 7.1238 GiB/s]
                 change:
                        time:   [-2.0332% -1.6904% -1.3140%] (p = 0.00 < 0.05)
                        thrpt:  [+1.3315% +1.7195% +2.0754%]
                        Performance has improved.
```

Reply:
Hmm, I had 5Gb/s per second before on master, but yes flex_master (3,6GB/s) is slower than the current.
I'm not sure that flex_master is the same as having the crate locally with regards to inlining?
I changed the check to:
```   
if start as usize + match_length + 16 - 1 > *output_ptr as usize
{
```
which led to a speedup from 4Gb/s to 4,7Gb/s, which is weird, it should not have that much of an effect. I changed some of the benchmark code and cannot reproduce it anymore (it's 4,4Gb/s now). So I guess it is related to some LTO stuff in the benchmark code.

Thanks for double checking.

Reply:
Yeah I find it very disturbing that the performance can vary so much due to  seemingly unrelated stuff 😩

Reply:
https://www.youtube.com/watch?v=r-TLSBdHe1A&t=642s this talk is addresses this issue of random memory layout changes affecting performance